#16888: Fix Conv2D when output is in Row Major

sankarmanoj-tt opened the pull request:
### Ticket
#16888 

### Problem description
Lots of test cases fail when output is in Row Major. 

### What's changed
Matmul_partials_cb is the input CB to the untilize block. Matmul_partials_cb may be directly mapped to the output buffer, or have its own memory allocated. 

If it's mapped to the output buffer, `use_partials_for_out == True`, then the read & write pointers have to be incremented after every in0 height block.

If it has it's own memory allocated,   `use_partials_for_out == False`, then the read & write pointers have to be reset to the original position.

### Checklist
- [x] Post commit CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/13467167167)
- [x] Blackhole Post commit (if applicable) [passes](https://github.com/tenstorrent/tt-metal/actions/runs/13565830704)
- [x] Model regression CI testing [passes](https://github.com/tenstorrent/tt-metal/actions/runs/13565834806) 
- [ ] Device performance regression CI testing passes (if applicable)
- [x] Nightly L2 [passes](https://github.com/tenstorrent/tt-metal/actions/runs/13565844287)
- [ ] **(For models and ops writers)** Full [new models](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) tests passes
- [ ] New/Existing tests provide coverage for changes

sankarmanoj-tt committed:
Fix Untilize Out

mywoodstock commented on tests/ttnn/unit_tests/operations/test_new_conv2d.py:
```diff
@@ -42,7 +42,7 @@ def run_conv(
     fp32_accum=False,
     packer_l1_acc=False,
-    output_layout=ttnn.TILE_LAYOUT,
+    output_layout=ttnn.ROW_MAJOR_LAYOUT,
```
changing the default? we also need to test the TILE output layout more extensively.

sankarmanoj-tt replied:
Most tests use the default value, and to ensure that untilize_out works, I changed the default.

pavlejosipovic replied:
I would keep tile as default as that is what is being used by models, add row major variant in tests where needed

sankarmanoj-tt replied:
Done.

mywoodstock commented on tests/ttnn/unit_tests/operations/test_new_conv2d.py:
```diff
@@ -115,6 +117,13 @@ def run_conv(
+    if (
+        shard_layout == ttnn.TensorMemoryLayout.HEIGHT_SHARDED
+        and output_channels > 256
+        and output_layout == ttnn.ROW_MAJOR_LAYOUT
```
Does this case work with TILE layout?

sankarmanoj-tt replied:
Yes.

pavlejosipovic replied:
This needs to added to validate function as well as TT_FATAL

pavlejosipovic replied:
bump

sankarmanoj-tt replied:
Done

mywoodstock commented on tests/ttnn/unit_tests/operations/test_new_conv2d.py:
```diff
@@ -115,6 +117,13 @@ def run_conv(
+        and output_channels > 256
+        and output_layout == ttnn.ROW_MAJOR_LAYOUT
+    ):
+        pytest.skip("Skipping when out_block_w > 8 for untilize_out == True")
```
also create a ticket to add support for this case, and mention the ticket number here.

sankarmanoj-tt replied:
https://github.com/tenstorrent/tt-metal/issues/18633

mywoodstock commented on tests/ttnn/unit_tests/operations/test_new_conv2d.py:
```diff
@@ -136,6 +145,11 @@ def run_conv(
+        if fp32_accum and packer_l1_acc and output_layout == ttnn.ROW_MAJOR_LAYOUT:
+            conv_config.output_layout = ttnn.TILE_LAYOUT
+            logger.warning(
+                "Forcing output_layout to TILE when act_block_h_override, fp32_accum and packer_l1_acc are enabled"
```
why TILE only for fp32_accum and packer_l1_acc?

sankarmanoj-tt replied:
It fails when untilize_out packer_l1 and fp32_accum are all enabled.

pavlejosipovic replied:
Can we add to validate function TT_FATAL for combinations we know are not working?

pavlejosipovic replied:
bump

sankarmanoj-tt replied:
Done.

mywoodstock commented on tests/ttnn/unit_tests/operations/test_new_conv2d.py:
```diff
@@ -169,7 +182,8 @@ def run_conv(
     )
-
+    # import numpy as np
+    # np.save("ref.npy",weights_device.cpu().to_torch().numpy())
```
delete

mywoodstock commented on tests/ttnn/unit_tests/operations/test_new_conv2d.py:
```diff
@@ -1213,7 +1229,7 @@ def test_resnet50_conv_wh_fp32(
 @pytest.mark.parametrize(
     "activations_dtype",
-    [ttnn.bfloat8_b],
+    [ttnn.bfloat16],
```
why change?

sankarmanoj-tt replied:
To test row major output.

mywoodstock commented on ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp:
```diff
@@ -86,7 +86,7 @@ bool check_non_tile_mul_width(
     bool is_non_tile_mul_width =
         (conv_config.shard_layout.has_value() && conv_config.shard_layout == TensorMemoryLayout::BLOCK_SHARDED) &&
-        conv_config.act_block_h_override == 0 &&
+        // conv_config.act_block_h_override == 0 &&
```
delete

mywoodstock commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/conv_bmm_tilize_col_major_out_blocks.cpp:
```diff
@@ -24,6 +24,9 @@
 // SliceRange src = SliceRange{.h0 = 0, .h1 = 32, .hs = 1, .w0 = 0, .w1 = 1, .ws = 1};
+// SliceRange row_range = SliceRange{.h0 = 0, .h1 = 1, .hs = 1, .w0 = 0, .w1 = 16, .ws = 1};
+// SliceRange col_range = SliceRange{.h0 = 0, .h1 = 16, .hs = 1, .w0 = 0, .w1 = 1, .ws = 1};
+// SliceRange sq_range = SliceRange{.h0 = 0, .h1 = 4, .hs = 1, .w0 = 0, .w1 = 4, .ws = 1};
```
lets delete the whole commented out block

sankarmanoj-tt replied:
Sure. I will remove once I've got all the tests passing.

sankarmanoj-tt replied:
Removed.

pavlejosipovic reviewed (commented):
Can you describe in PR what was the issue in compute kernel and what was done to fix, I can't follow the code.

sankarmanoj-tt committed:
#0: Rebase fix

#0: Rebase fix

#0: Fix tests

#0: Fix tests

#0: Removed comments

#0: Make partials_use_out as a constexpr

#0: Revert default layout to TILE

#0: Revert to physical sharding

#0: Disable failing tests

#0: Cleanup

#0: Add comment

pavlejosipovic reviewed (commented):
What is the impact of this change on tests runtime?

pavlejosipovic commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/conv_bmm_tilize_col_major_out_blocks.cpp:
```diff
@@ -135,6 +130,10 @@ void MAIN {
     SFPU_OP_INIT_ACTIVATION
 #endif
+    UNPACK(uint32_t partials_cb_read_ptr = get_local_cb_interface(matmul_partials_cb).fifo_rd_ptr;
+           const bool use_partials_for_out = (partials_cb_read_ptr == get_local_cb_interface(out_cb_id).fifo_rd_ptr);)
```
We should know form factory if this the case and we could pass this in as constexpr arg, and make all the if's if constexpr?

sankarmanoj-tt replied:
Done.

mywoodstock commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/conv_bmm_tilize_col_major_out_blocks.cpp:
```diff
@@ -313,7 +315,14 @@ void MAIN {
-
+                if (curr_matmul_out_cb == matmul_partials_cb) {
+                    UNPACK(
+                        if (!use_partials_for_out) get_local_cb_interface(matmul_partials_cb).fifo_rd_ptr =
```
make `constexpr`

sankarmanoj-tt replied:
curr_matmul_out_cb is not a const value.

sankarmanoj-tt replied:
Done.

mywoodstock commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/conv_bmm_tilize_col_major_out_blocks.cpp:
```diff
@@ -313,7 +315,14 @@ void MAIN {
+                        if (!use_partials_for_out) get_local_cb_interface(matmul_partials_cb).fifo_rd_ptr =
+                            partials_cb_read_ptr);
+                    PACK(
+                        if (!use_partials_for_out) get_local_cb_interface(matmul_partials_cb).fifo_wr_ptr =
```
make `constexpr`

sankarmanoj-tt replied:
Done.

mywoodstock commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/conv_bmm_tilize_col_major_out_blocks.cpp:
```diff
@@ -366,7 +375,9 @@ void MAIN {
             if constexpr (matmul_partials_cb == mm_out_cb_id) {
-                UNPACK(get_local_cb_interface(matmul_partials_cb).fifo_rd_ptr = partials_cb_read_ptr);
+                UNPACK(
+                    if (use_partials_for_out) get_local_cb_interface(matmul_partials_cb).fifo_rd_ptr =
```
make `constexpr`

sankarmanoj-tt replied:
Done.

mywoodstock commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/conv_bmm_tilize_col_major_out_blocks.cpp:
```diff
@@ -416,6 +427,10 @@ void MAIN {
                     in1_index_subblock_offset += out_subblock_w;
                 }  // for in1_num_subblocks
             }  // in0_num_subblocks
+            if (untilize_out) {
```
make `constexpr`

sankarmanoj-tt replied:
Done.

sankarmanoj-tt commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/conv_bmm_tilize_col_major_out_blocks.cpp:
```diff
@@ -313,7 +315,14 @@ void MAIN {
                     in0_index_subblock_offset += in0_subblock_num_tiles;
                 }
-
+                if (curr_matmul_out_cb == matmul_partials_cb) {
```
curr_matmul_out_cb is not a const value.

pavlejosipovic commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp:
```diff
@@ -226,17 +226,26 @@ std::vector<TensorSpec> OptimizedConvNew::compute_output_specs(const std::vector
         } else if (this->memory_config.memory_layout == TensorMemoryLayout::WIDTH_SHARDED) {
             uint32_t total_height_tiles = padded_output_shape.volume() / padded_output_shape[-1] / TILE_HEIGHT;
-            std::array<uint32_t, 2> shard_shape = {
+
```
I think it's tricky to use logical sharding for conv output only for width sharding.
In general most ops are not ready to consume logical sharding (like TMs) and change like this could trigger models to fail.
Now people will have to have in mind that widht sharing outputs logical shards but HS and BS output physical shard.

What was to reason to go for logical sharding here?

sankarmanoj-tt replied:
With width sharding and the output as Row Major Layout, [this](https://github.com/tenstorrent/tt-metal/blob/45186bcf465d6bad3d5e8efaf3ef5afb9e2de2d9/ttnn/cpp/ttnn/tensor/tensor_spec.cpp#L44) condition was failing when the output height * width was not a multiple of 32. 

When the output was in Tile Layout, it would correctly pad the shard. However, in Row Major Layout, the physical size of the shard would always be the logical size.

sankarmanoj-tt replied:
Should HS and BS also be changed to Logical Sharding?

sankarmanoj-tt replied:
The initial error was specific to WS. https://github.com/tenstorrent/tt-metal/blob/45186bcf465d6bad3d5e8efaf3ef5afb9e2de2d9/ttnn/cpp/ttnn/tensor/tensor_spec.cpp#L44

I can port the other types to logical sharding too.

pavlejosipovic replied:
Can you try to hit the same problem with other shardings?

sankarmanoj-tt commented:
There is around a 5% increase in kernel execution time when the output is in Row Major. 
[ops_perf_results_rm_comp_2025_02_27_07_57_07.csv](https://github.com/user-attachments/files/19004858/ops_perf_results_rm_comp_2025_02_27_07_57_07.csv)

pavlejosipovic commented:
> There is around a 5% increase in kernel execution time when the output is in Row Major. [ops_perf_results_rm_comp_2025_02_27_07_57_07.csv](https://github.com/user-attachments/files/19004858/ops_perf_results_rm_comp_2025_02_27_07_57_07.csv)

Sorry I wasn't clear enough I asked about is runtime of test_new_conv2d.py increased now, and if so by how much (as we were trying to reduce the runtime of this to make post commit faster)

pavlejosipovic commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp:
```diff
@@ -210,48 +224,26 @@ std::vector<TensorSpec> OptimizedConvNew::compute_output_specs(const std::vector
-        } else {
-            TT_THROW("Unsupported shard scheme");
-        }
+        CoreRangeSet shard_grid;
```
Do you have all relevant pipelines passing with this version of the code?
If yes this is great!

If we stick to this please document this chagne that conv2d has switched from physical to logical sharding for the output tensor.

sankarmanoj-tt replied:
I've removed logical sharding for the time being. There is a some work to be done to get the models passing, and I don't want to do all of that in one PR.

pavlejosipovic commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp:
```diff
@@ -210,48 +224,26 @@ std::vector<TensorSpec> OptimizedConvNew::compute_output_specs(const std::vector
+        auto shard_shape = physical_shard_shape;
+
+        auto shard_spec = ShardSpec{
+            shard_grid, shard_shape, physical_shard_shape, this->memory_config.shard_spec.value().orientation};
```
Should we use different shapes for physical and logical shard?

pavlejosipovic commented on ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_sharded_program_factory.cpp:
```diff
@@ -1123,6 +1125,21 @@ operation::ProgramWithCallbacks multi_core_optimized_conv_sharded_v2_impl(
+        partials_cb_uses_output = cb_config_matmul_partials.globally_allocated_address().value() ==
+                                  cb_config_output.globally_allocated_address().value();
+    }
+    log_info(
```
did you intend to use log_debug here?

pavlejosipovic commented:
> What is the impact of this change on tests runtime?

BUMP

pavlejosipovic commented on models/experimental/functional_unet/tt/unet_shallow_ttnn.py:
```diff
@@ -311,6 +311,9 @@ def __call__(self, x, residual):
 
         x = self.upsample(x)
 
+        residual = ttnn.sharded_to_interleaved(residual, ttnn.L1_MEMORY_CONFIG)
```
what is the perf impact of this?

pavlejosipovic commented on tt_metal/impl/buffers/buffer.cpp:
```diff
@@ -604,8 +605,47 @@ const std::shared_ptr<const BufferPageMapping>& Buffer::get_buffer_page_mapping(
 
-bool ShardSpec::operator==(const ShardSpec&) const = default;
-bool ShardSpec::operator!=(const ShardSpec&) const = default;
+bool ShardSpec::operator==(const ShardSpec& other) const {
```
What is the motivation for this?

sankarmanoj-tt replied:
Lots of ops were comparing the shard spec of the inputs, and failing if they weren't equal. There were scenarios where both inputs were exactly the same, except one used logical sharding and the other physical sharding.

sankarmanoj-tt commented:
@pavlejosipovic 
test_new_conv2d.py stats : 791 passed, 350 skipped, 5 xfailed, 894 warnings, 24 errors in 559.48s (0:09:19)

sankarmanoj-tt commented on tests/ttnn/unit_tests/operations/test_new_conv2d.py:
```diff
@@ -127,6 +129,14 @@ def run_conv(
+        and output_layout == ttnn.ROW_MAJOR_LAYOUT
+    ):
+        output_layout = ttnn.TILE_LAYOUT
+        logger.warning("Disabling untilize_out when out_c > 256 for Height Sharded")
```
Done.

pavlejosipovic reviewed (commented):
> @pavlejosipovic 
> test_new_conv2d.py stats : 791 passed, 350 skipped, 5 xfailed, 894 warnings, 24 errors in 559.48s (0:09:19)
> 
> The time has increased due to adding two different output layouts for each test. 

I don't know what is the baseline on the machine you got this measurements from, but If this change is doubling run time of test_new_conv2d.py  that is not OK, as we have broather push to reduce runtime on post-commit.
Not testing features is also not OK, so we need to move things nightly pipeline.
You can move/add tests to this path.
tests/ttnn/nightly/unit_tests/operations/conv2d/...

sankarmanoj-tt committed:
#0: Rebase fix

sankarmanoj-tt committed:
#0: Revert conv2d UTs

sankarmanoj-tt committed:
#0: Updated conv2d UTs

sankarmanoj-tt committed:
#0: Rebase fix
